src/OpenTelemetry/Metrics/MetricExporterExtensions.cs

```diff
@@ -0,0 +1,69 @@
+// <copyright file="MetricExporterExtensions.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System;
+using System.Threading;
+
+namespace OpenTelemetry.Metrics
+{
+    /// <summary>
+    /// Contains extension methods for the <see cref="BaseExporter{Metric}"/> class.
+    /// </summary>
+    public static class MetricExporterExtensions
+    {
+        /// <summary>
+        /// Attempts to collect the metrics, blocks the current thread until
+        /// metrics collection completed or timed out.
+        /// </summary>
+        /// <param name="exporter">BaseExporter{Metric} instance on which Collect will be called.</param>
+        /// <param name="reader">BaseExportingMetricReader instance on which Collect will be called.</param>
+        /// <param name="timeoutMilliseconds">
+        /// The number (non-negative) of milliseconds to wait, or
+        /// <c>Timeout.Infinite</c> to wait indefinitely.
+        /// </param>
+        /// <returns>
+        /// Returns <c>true</c> when metrics collection succeeded; otherwise, <c>false</c>.
+        /// </returns>
+        /// <exception cref="System.ArgumentOutOfRangeException">
+        /// Thrown when the <c>timeoutMilliseconds</c> is smaller than -1.
+        /// </exception>
+        /// <remarks>
+        /// This function guarantees thread-safety.
+        /// </remarks>
+        public static bool Collect(this BaseExporter<Metric> exporter, BaseExportingMetricReader reader, int timeoutMilliseconds = Timeout.Infinite)
```

Comment:
I was chatting to @reyang about this a bit on Slack. Nothing against the extension, just not sure how Exporter will get the second param (MetricReader). Also seems a bit odd the extension is on Exporter, but really only Reader is used.

A couple of ideas for exposing MetricReader to exporter...

1. Don't. Just feed it to Prometheus where we need it. If some other exporter comes along that needs to call collect, we tackle it then

```csharp
public PrometheusExporter(PrometheusExporterOptions options, MetricReader parentMetricReader)
```

2. I guess @alanwest tried to do some base classes but ran into issues with `InMemoryExporter<T>`. What if we went with an interface? Closest thing C# has to composition.

```csharp
public interface ICollectingExporter
{
   MetricReader ParentMetricReader { get; set; } 
}

public class PrometheusExporter : ICollectingExporter
{
   public MetricReader ParentMetricReader { get; set; } 
}
```

During MeterProvider build-up we then just do an `is` check and set the `ParentMetricReader` on any exporter that implements the interface (basically if exporter asks for it).

Reply:
Another idea we could explore:

3. leverage Event https://docs.microsoft.com/en-us/dotnet/standard/events/

Reply:
_Please take my comment from a distance. Just learning the API. But maybe some fresh eyes will help somehow._

This extension method design is strange. Does it mean that the `BaseExporter<Metric>` has to be coupled with `BaseExportingMetricReader`? If so then maybe it will be handy to create `BaseMetricExporter`?

Regarding @CodeBlanch I definitely prefer approach 1 (composition) as it leads to less coupling. Also, I consider interface getters and setters as a code smell.

Not sure how we would like to use `event` but personally I prefer `IObservable<T>` as it is easier to compose, works better with extension methods etc.

Reply:
Option 1 is on the bottom of my list in terms of preference, I even suspect if we could do that (reader.ctor needs exporter, how could exporter.ctor take reader?). I guess we might be able to tweak it and make it work, but I have strong feeling that the `exporter.ctor(options, reader)` is a bad design:
1. why would reader be a separate parameter, rather than part of the options.
2. why would we have the `options, reader` ordering instead of `reader, options = optional`.
3. what if we later we figured that we need to add another parameter?

Reply:
I've taken option 2 from @CodeBlanch.

A basic pull-only exporter would look like this:

```csharp
[ExportModes(ExportModes.Pull)]
private class PullOnlyMetricExporter : BaseExporter<Metric>, IPullMetricExporter
{
    private Func<int, bool> funcCollect;

    public Func<int, bool> Collect
    {
        get => this.funcCollect;
        set { this.funcCollect = value; }
    }

    public override ExportResult Export(in Batch<Metric> batch)
    {
        return ExportResult.Success;
    }
}
```

And the exporter can trigger Collect by using `exporter.Collect(timeoutMilliseconds)`. Any direct invocation on `reader.Collect` and `provider.ForceFlush` would result in `false`.

```csharp
// in the exporter code where we handle scraper
exporter.Collect(timeoutMilliseconds);
```

Reply:
@CodeBlanch @pellared PTAL.

Reply:
Looks OK. I will create a PR, if I come up with anything potentially better.